### ISSUE TITLE
Add structure for licensing, possibly validated using Reuse or a similar standard?

### ISSUE THREAD

**Issue by ferdnyc** [jenstroeger/python-package-template#21]
Interesting project, I think you've got a pretty good start here, but one thing that strikes me as lacking is a repo-wide view of license management.

Clearly establishing license terms is so important in open-source projects (both for the author(s) of the code, and for anyone who's considering making use of it in their own projects), yet it's often handled haphazardly as an afterthought — or, just as often, not handled _at all_ until someone pesters the maintainer into [acting out of frustration](https://github.com/pjreddie/darknet/commit/e6b3006d1b67f97be70398436fd6cce919897015). Typically that kind of thing causes problems down the line for _someone_ (probably **not** the originator of the... "bespoke" license in question).

Adopting a standard for licensing right out of the gate is a simpler, saner solution. And it can be as simple as committing to standard practices defined by something like [the Reuse project](https://reuse.software/). Just like most projects will commit to following semantic versioning practices, adopting reuse-verified license compliance requires nothing more than a commitment to put in the work.

**Comment by ferdnyc** [comment]
> Thank you, @ferdnyc, that makes a lot of sense!
> 
> I wasn’t aware of the Reuse Project or its compliance tools, 

Yeah, I just learned of it a month or two ago. But they seem to have a good and level-headed take on this stuff.

It can definitely _seem_ a bit off and pointless at first, having to deal with adding license tags to things like a `.gitignore` file or a workflow for a GitHub bot. Occasionally the added text for licensing purposes is _longer_ than the functional content of the file! For someone who just doesn't see the point in any of it, I'm sure the initial setup work to bring a repo compliant would feel like a bunch of tedious busywork. 

(Though it's certainly a lot easier to start off that way, than to have to retroactively figure all this stuff out for existing code. ...Speaking most definitely from experience.)

But as I noted earlier, there are a lot of parallels to semantic versioning/releasing: It's not a thing _being imposed on_ a project, it's an ideal that the project's developers may _choose to commit to_, if they see the benefits as outweighing any additional restrictions or effort involved.

> and after browsing through their website I’d be more than happy to integrate their tools (e.g. the [pre-commit hook](https://reuse.software/dev/#pre-commit-hook)).

A pre-commit hook is certainly an option, for developers who want to keep themselves honest that way. I've always been a bit wary of Git hooks because they don't always play well with everyone's preferred mode of interacting with git. (Some UIs and tools have been fully integrated with all aspects of the git model, they gracefully recover from rejected commit attempts and notify the user in a clear and actionable way. Others... not so much.)

More interesting, in my view, are the various CI tools they offer, their [Marketplace GitHub Action](https://github.com/marketplace/actions/reuse-compliance-check) being the most obviously useful in this context. A CI check that flags any licensing issues and blocks PR merges is a no-brainer for a project that wants to stay Reuse-compliant.

(Won't prevent project committers from potentially pushing code with issues, if they have direct commit access to a branch... but it would mean that PR checks _start_ failing after they do, through no fault of the PR submitter. :roll_eyes: )

> Would you like to open a PR?

Sure, I can do that. What do I need to know about the _current_ contents of the repo, license-wise? Are there any files from external sources that have to be treated specially, or is it all original work covered by the blanket MIT license?

**Comment by ferdnyc** [comment]
(I know I saw `sphinx-quickstart`output in one file, from a quick scan. AIUI files generated that way — especially as it's an interactive tool — are considered fully original works; you're free to license it however you choose, so no issues there.)

**Comment by ferdnyc** [comment]
@behnazh  Interesting! Reuse and the license checkers you mentioned actually have completely different aims, I would think that they're broadly compatible (though, really, they wouldn't be touching any of the same things so maybe "complementary" is a better word.)

The focus with Reuse is on making _your own_ code ready to be reused, by making sure all local files are tagged with a valid license identifier. It doesn't do anything with dependencies at all.

(Reuse also avoids trying to boil an entire package down to a single license, because as code accumulates and files are pulled in from outside sources, the fact is that a single repo can contain files under a mix of 2, 3, 5, who-knows-how-many different licenses. (If a font is bundled, that's often a different license. Bundled icons, same thing. If there are data files included with the unit tests, those very often don't  have any clear licensing _at all_. To assume they're licensed under the same terms as the overall project is often just wishful thinking.

That being said, it would be an interesting extension to reuse, if there were a tool that could pull in dependency licenses and add that info to the local license metadata. Any additional license texts could even be added to the local `LICENSES/` directory alongside the local ones.

...BTW, am I the only one who finds it odd that `pip-license-checker` is written in Clojure?

**Comment by jenstroeger** [comment]
Thank you all for your comments, I really appreciate that! It looks to me like the original thread has split into three somewhat independent facets:

 1. Formalize and clearly state/implement license information for this repository, e.g. using the [Reuse Project](https://reuse.software/) like @ferdnyc proposed initially.
 2. Check the licenses of all dependent packages, e.g. using @pilosus’s [pip-license-checker](https://github.com/pilosus/action-pip-license-checker) or [Black Duck](https://www.blackducksoftware.com/) like @behnazh suggests. In both cases, it might be useful if dependent licenses would be output as [SPDX identifiers](https://spdx.org/licenses/) because… next point.
 3. Cross-reference the licenses of all dependent packages with the license of _this_ package to check for conflicting uses.

I think point 1. was discussed [above]:

> @ferdnyc Sure, I can do that. What do I need to know about the current contents of the repo, license-wise? Are there any files from external sources that have to be treated specially, or is it all original work covered by the blanket MIT license?

I think currently all files in the repo can be covered by the example blanket MIT license. But I also think it’d make sense to add a section to the README on how to move forward managing license information — wrt. all three points.

As for point 2. we could probably add the [license checker action](https://github.com/pilosus/action-pip-license-checker) and it looks like the checker’s `fail` option would address point 3. Although I wonder if the two could be knitted closer together 🤔

**Comment by jenstroeger** [comment]
In this context, [PEP 639](https://www.python.org/dev/peps/pep-0639/) is currently being drafted and discussed here: https://github.com/python/peps/pull/2164 (see [this](https://discuss.python.org/t/pep-639-improving-license-clarity-with-better-package-metadata/2154) initial forum thread that spawned the PEP).

I haven’t read all of it yet, though, but probably worth taking into consideration here.

**Comment by CAM-Gerlach** [comment]
Thanks! Let us know if you have any questions or feedback. The scope of that PEP has been limited to project-wide license expressions as opposed to formally specifying per-file convention, though the standard `SPDX-License-Expression` header is referred to, but 

As a general bit of feedback, for a template whose stated focus is on PEP 518 (which would tend to imply PEP 517), its a little surprising to see no `build-system` section in the `pyproject.toml` nor and all the build tool configuration and boilerplate code in `setup.py` instead of simpler, declarative configuration in `setup.cfg`.

Also, maybe worth considering cookiecutter (or a similar tool) to save a large amount of user effort and potential errors manually replacing everything, which would be particularly important when it comes to standardized license headers, copyright lines, license metadata, etc.

**Comment by jenstroeger** [comment]
Related discussion: [PEP 639, Round 2: Improving license clarity with better package metadata](https://discuss.python.org/t/pep-639-round-2-improving-license-clarity-with-better-package-metadata/12622)

**Comment by CAM-Gerlach** [comment]
FWIW, I hope to _finally_ do what is hopefully the last big PEP update in the next couple days, after being busy with my research, work and other FOSS things.

**Comment by jenstroeger** [comment]
@acrobat888 suggested to look at [LicenseCheck](https://github.com/FHPythonUtils/LicenseCheck) and [pip-licenses](https://github.com/raimon49/pip-licenses) as well.

**Comment by jenstroeger** [comment]
And see also [PEP 639](https://peps.python.org/pep-0639/) for related discussion.

**Comment by jenstroeger** [comment]
Another interesting package is [license-expression](https://github.com/aboutcode-org/license-expression).
